src/notation/view/pianokeyboard/pianokeyboardview.cpp

```diff
@@ -0,0 +1,476 @@
+
+            qreal center = rect.width() / 2;
+
+            topPiecePath.moveTo(left, top);
+            topPiecePath.lineTo(left, bottom - cornerRadius);
+            topPiecePath.quadTo(left, bottom, center, bottom);
+            topPiecePath.quadTo(right, bottom, right, bottom - cornerRadius);
+            topPiecePath.lineTo(right, top);
+            topPiecePath.closeSubpath();
+
+            topPieceGradient.setColorAt(0.0, backgroundColor);
+            topPieceGradient.setStart(0.0, top);
+            topPieceGradient.setFinalStop(0.0, bottom);
+
+            // Make bottom piece
+            top = rect.height() - m_spacing - bottomPieceHeight;
+            bottom = rect.height() - m_spacing;
+
+            bottomPiecePath.moveTo(left, top + cornerRadius);
+            bottomPiecePath.quadTo(left, top, center, top);
+            bottomPiecePath.quadTo(right, top, right, top + cornerRadius);
+            bottomPiecePath.lineTo(right, bottom);
+            bottomPiecePath.lineTo(left, bottom);
+            bottomPiecePath.closeSubpath();
+
+            bottomPieceGradient.setColorAt(1.0, backgroundColor);
+            bottomPieceGradient.setStart(0.0, top);
+            bottomPieceGradient.setFinalStop(0.0, bottom);
+        }
+
+        QColor highlightColor = m_blackKeyStateColors[m_controller->keyState(key)];
+        topPieceGradient.setColorAt(1.0, highlightColor);
+        bottomPieceGradient.setColorAt(0.0, highlightColor);
+
+        painter->translate(rect.topLeft());
+        painter->fillRect(backgroundRect, backgroundColor);
+        painter->fillPath(topPiecePath, topPieceGradient);
+        painter->fillPath(bottomPiecePath, bottomPieceGradient);
+        painter->translate(-rect.topLeft());
+    }
+}
+
+int PianoKeyboardView::numberOfKeys() const
+{
+    return m_numberOfKeys;
+}
+
+void PianoKeyboardView::setNumberOfKeys(int number)
+{
+    if (m_numberOfKeys == number) {
+        return;
+    }
+
+    switch (number) {
+    case 128:
+        m_lowestKey = 0;
+        m_numberOfKeys = 128;
+        break;
+    case 88:
+        m_lowestKey = 21;
+        m_numberOfKeys = 88;
+        break;
+    case 61:
+        m_lowestKey = 36;
+        m_numberOfKeys = 61;
+        break;
+    case 49:
+        m_lowestKey = 36;
+        m_numberOfKeys = 49;
+        break;
+    case 25:
+        m_lowestKey = 48;
+        m_numberOfKeys = 25;
+        break;
+    default:
+        return;
+    }
+
+    emit numberOfKeysChanged();
+    calculateKeyRects();
+    update();
+}
+
+void PianoKeyboardView::moveCanvas(qreal dx)
+{
+    m_scrollOffset += dx;
+    adjustKeysAreaPosition();
+    update();
+}
+
+qreal PianoKeyboardView::keyWidthScaling() const
+{
+    return m_keyWidthScaling;
+}
+
+void PianoKeyboardView::scale(qreal factor, qreal x)
+{
+    setScaling(m_keyWidthScaling * factor, x);
+}
+
+void PianoKeyboardView::setScaling(qreal scaling, qreal x)
+{
+    qreal newScaling = std::clamp(scaling, 0.5, 2.0);
```

Comment:
Perhaps these numbers are from the PianoKeyboardPanelContextMenuModel? You should use static constants here instead of raw numbers